File: solutions/python/run-length-encoding/1/run_length_encoding.py
from itertools import groupby


def decode(code):
    count = ""
    decoded = ""
    for group_count, group in groupby(code):
        if group_count.isnumeric():
            count += str(group_count)
        elif not count:
            decoded += str(group_count)
        else:
            decoded += str(int(count) * group_count)
            count = ""
    return decoded


def encode(s):
    code = ""
    for group_count, group in groupby(s):
        section = sum(1 for _ in group)
        if section > 1:
            code += str(section) + group_count
        else:
            code += group_count
    return code

File: solutions/kotlin/run-length-encoding/1/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String {
        if (input.isEmpty()) return ""
    
        val encoded = StringBuilder()
        var count = 1
    
        for (i in 1 until input.length) {
            if (input[i] == input[i - 1]) {
                count++ // Increment count if the current character is the same as the previous
            } else {
                // Append count and character only if count is greater than 1
                if (count > 1) encoded.append(count)
                encoded.append(input[i - 1]) // Append the character
                count = 1 // Reset count
            }
        }
    
        // Append the last character and count
        if (count > 1) encoded.append(count)
        encoded.append(input.last())
    
        return encoded.toString()
    }


    fun decode(encoded: String): String {
        val decoded = StringBuilder()
        var count = 0
    
        for (char in encoded) {
            if (char.isDigit()) {
                // Build the count from consecutive digits
                count = count * 10 + (char - '0')
            } else {
                // Append the character `count` times, or once if count is zero
                decoded.append(char.toString().repeat(if (count > 0) count else 1))
                count = 0 // Reset count after using it
            }
        }
    
        return decoded.toString()
    }

}
